### ISSUE TITLE
Test best visualisation methods

### ISSUE THREAD

**Issue by Cryaaa** [BiAPoL/napari-clusters-plotter#97]
In order to find out which method is fastest for visualising cluster identities we should have some testing notebooks which compare 

- pyclesperanto cluster image generation
- numpy cluster image generation (possibly numba accelerated? 
- use new colormap for cluster visualisation (for timelapse connect timepoint changed event to selection of new colormap)

Should be done before #81 since we could then decide which method to implement and if it makes sense to keep pyclesperanto at all

**Comment by Cryaaa** [comment]
As you guys can see the custom colormaps are pretty much head to head with pyclesperanto when viewing in 3D but are far superior when viewing in 2D. Since the creation of these does not require any extra libraries I think we should revert back to this way of visualising clusters. (I have figured out how to do that now).

**Comment by zoccoler** [comment]
Hi Ryan @Cryaaa ,

That sounds reasonable to me looking at these results. Is the code for these tests available somewhere? I would love to take a look! 

Best,
Marcelo

**Comment by jo-mueller** [comment]
Hey @Cryaaa,

just to see whether I got this straight - you compared how long it took to create the label images with the label color according to the assigned cluster index? How many objects did you color in the plot above? Or in other words - is the difference so significant for images with 100-1000 objects or would I have to display ten-thousands of objects to find the differences above?

**Comment by Cryaaa** [comment]
@zoccoler, @haesleinhuepf, @jo-mueller   
> Is the code for these tests available somewhere?
> I also would like to see the benchmarking code btw. :-)

I will upload it with a new PR soon. I can tag you in where I change the visualisation method! 

> just to see whether I got this straight - you compared how long it took to create the label images with the label color according to the assigned cluster index? How many objects did you color in the plot above? Or in other words - is the difference so significant for images with 100-1000 objects or would I have to display ten-thousands of objects to find the differences above?

So for the 3D dataset the embryo had roughly 3000 labels for which the cluster image was generated and the 4D dataset had roughly 3000 cells per frame and 32 frames (although only single time slices matter since the generation of label images is handled by dask meaning only one has to be generated for visualisation). The cluster image generation with CLE and numpy actually creates a new label image with the labels corresponding to cluster identity wheres the mapping is just a colormap that maps a label to its cluster identity color.

> Thus, if you go down the custom colormap road, please make sure that the functionality is not lost. We might need the actual cluster-id-label-images.

I definitely agree! It's a cool functionality but I don't think people would automatically think about it and I would rather have faster image generation if possible to be honest. We can add a function though which generates a cluster image with labels corresponding to cluster identity for exporting to disc (maybe even have a simple magicgui widget for it).

**Comment by haesleinhuepf** [comment]
An idea for improvement:

> The cluster image generation with CLE [...] actually creates a new label image with the labels corresponding to cluster identity wheres the mapping is just a colormap that maps a label to its cluster identity color.

If you could reuse the pre-existing label image (when updating the visualization, at the second and following updates), the compute time for writing values into the label image should go down to almost 0 when using clesperanto. Read more: https://github.com/clij/clij-docs/blob/master/faq.md#does-reusing-memory-bring-additional-speed-up

**Comment by Cryaaa** [comment]
> 

@haesleinhuepf: Ahhhh okay so if I were not pulling the image and deleting the GPU image it would be much faster if I understood it right? To do that the GPU image would have to live outside of the function though am I right? I guess since we want to move away from pyclesperanto anyway it might make sense to remove the code entirely though

**Comment by Cryaaa** [comment]
As discussed in #100 we can't speed up 3D visualisation anyway and since the cluster image is not necessarily the bottleneck we will keep the old way of generating the cluster image.
